src/main/java/aquality/selenium/core/elements/CachedElementStateProvider.java

```diff
@@ -0,0 +1,122 @@
+package aquality.selenium.core.elements;
+
+import aquality.selenium.core.elements.interfaces.IElementCacheHandler;
+import aquality.selenium.core.logging.Logger;
+import aquality.selenium.core.waitings.IConditionalWait;
+import org.openqa.selenium.By;
+import org.openqa.selenium.NoSuchElementException;
+import org.openqa.selenium.StaleElementReferenceException;
+import org.openqa.selenium.WebElement;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeoutException;
+import java.util.function.BooleanSupplier;
+import java.util.function.Predicate;
+
+/**
+ * Provides functions to retrive the state for cached element.
+ */
+public class CachedElementStateProvider extends ElementStateProvider {
+
+    private final By locator;
+    private final IConditionalWait conditionalWait;
+    private final IElementCacheHandler elementCacheHandler;
+
+    public CachedElementStateProvider(By locator, IConditionalWait conditionalWait, IElementCacheHandler elementCacheHandler) {
+        this.locator = locator;
+        this.conditionalWait = conditionalWait;
+        this.elementCacheHandler = elementCacheHandler;
+    }
+
+    protected List<Class<? extends Exception>> getHandledExceptions() {
+        return Arrays.asList(StaleElementReferenceException.class, NoSuchElementException.class);
+    }
+
+    protected boolean tryInvokeFunction(Predicate<WebElement> predicate) {
+        return tryInvokeFunction(predicate, getHandledExceptions());
+    }
+
+    protected boolean tryInvokeFunction(Predicate<WebElement> predicate, List<Class<? extends Exception>> handledExceptions) {
+        try {
+            return predicate.test(elementCacheHandler.getElement(getZeroTImeout(), ElementState.EXISTS_IN_ANY_STATE));
+        } catch (Exception exception) {
+            if (handledExceptions.contains(exception.getClass())) {
+                return false;
+            }
+            throw exception;
+        }
+    }
+
+    protected boolean waitForCondition(BooleanSupplier condition, String conditionName, Long timeout) {
+        boolean result = conditionalWait.waitFor(condition, timeout, null);
+        if (!result) {
+            String timeoutString = timeout == null ? "" : String.format("of %1$s seconds", timeout);
+            String message = String.format(
+                    "Element %1$s has not become %2$s after timeout %3$s", locator, conditionName.toUpperCase(), timeoutString);
+            Logger.getInstance().warn(message);
```

Comment:
would you like to localize this message?

Reply:
fixed